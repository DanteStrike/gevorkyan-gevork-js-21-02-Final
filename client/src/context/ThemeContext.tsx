import React from 'react';
import {DomUtils} from '../utils';

export enum Theme {
  LIGHT = `light`,
  DARK = `dark`,
  DEFAULT = `light`,
}

export interface IThemeContext {
  theme: Theme;
  setTheme: (theme: Theme) => void;
}
interface IThemeProviderProps {
  children: React.ReactNode;
  storageKey: string;
}
interface IThemeProviderState {
  theme: Theme;
}

const ThemeContext = React.createContext<IThemeContext>({
  theme: Theme.LIGHT,
  setTheme: () => {},
});

class ThemeProvider extends React.PureComponent<IThemeProviderProps, IThemeProviderState> {
  constructor(props: IThemeProviderProps) {
    super(props);

    this.state = {
      theme: Theme.DEFAULT,
    };

    this.handleThemeChange = this.handleThemeChange.bind(this);
  }

  componentDidMount(): void {
    const {storageKey} = this.props;
    const storedTheme = (localStorage.getItem(storageKey) as Theme) || Theme.DEFAULT;

    this.setState({theme: storedTheme}, () => {
      DomUtils.addBodyClass(`theme--${storedTheme}`);
    });
  }

  handleThemeChange(value: Theme) {
    const {storageKey} = this.props;
    const {theme} = this.state;
    DomUtils.removeBodyClass(`theme--${theme}`);

    this.setState({theme: value}, () => {
      DomUtils.addBodyClass(`theme--${value}`);
    });
    localStorage.setItem(storageKey, value);
  }

  render() {
    const {theme} = this.state;
    const {children} = this.props;

    return <ThemeContext.Provider value={{theme, setTheme: this.handleThemeChange}}>{children}</ThemeContext.Provider>;
  }
}

export {ThemeContext, ThemeProvider};
