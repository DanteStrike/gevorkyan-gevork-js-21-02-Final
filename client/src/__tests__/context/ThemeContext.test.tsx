import {shallow} from 'enzyme';
import React from 'react';
import {IThemeContext, Theme, ThemeProvider} from '../../context/ThemeContext';

describe(`Component ThemeProvider should render correctly`, () => {
  it(`snapshot default test`, () => {
    const wrapper = shallow(
      <ThemeProvider storageKey="key">
        <div id="content" />
      </ThemeProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});

describe(`Component ThemeProvider should work correctly`, () => {
  it(`snapshot default test`, () => {
    const wrapper = shallow(
      <ThemeProvider storageKey="key">
        <div id="content" />
      </ThemeProvider>
    );
    let value = wrapper.prop(`value`) as IThemeContext;
    expect(value.theme).toBe(Theme.DEFAULT);
    value.setTheme(Theme.DARK);
    wrapper.update();
    value = wrapper.prop(`value`) as IThemeContext;
    expect(value.theme).toBe(Theme.DARK);
    value.setTheme(Theme.LIGHT);
    value = wrapper.prop(`value`) as IThemeContext;
    expect(value.theme).toBe(Theme.LIGHT);
  });
});
