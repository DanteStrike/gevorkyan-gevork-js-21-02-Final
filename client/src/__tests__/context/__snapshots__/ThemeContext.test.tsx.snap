// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component ThemeProvider should render correctly snapshot default test 1`] = `
<ContextProvider
  value={
    Object {
      "setTheme": [Function],
      "theme": "light",
    }
  }
>
  <div
    id="content"
  />
</ContextProvider>
`;
